File: phase1-data/src/main/java/edu/gemini/tac/persistence/queues/metrics/PartnerFairness.java
//
// $
//

package edu.gemini.tac.persistence.queues.metrics;

import edu.gemini.tac.persistence.Partner;

import java.util.Map;

// There is no need to store this, since we can compute it easily.


/**
 * Partner fairness calculation.  Measured as a percentage of error from
 * the partner allocation.  Can be calculated
 *
 * Percentages are used because partners have wildly differing time allocations.
 * The standard deviation of all remaining time percentages gives an indication
 * of how fair the algorithm has been.
 */
public final class PartnerFairness {

    private final double standardDeviation;
    private final Map<Partner, Double> percentErrorMap;

    public PartnerFairness(Map<Partner, Double> remaining, Map<Partner, Double> available) {
        percentErrorMap = MapOps.combine(remaining, MapOps.PERCENT, available);
        double meanErrorPercent = MapOps.avg(percentErrorMap);

        Map<Partner, Double> diff   = MapOps.apply(percentErrorMap, MapOps.SUBTRACT, meanErrorPercent);
        Map<Partner, Double> sqDiff = MapOps.apply(diff, MapOps.POW, 2.0);
        standardDeviation = Math.sqrt(MapOps.avg(sqDiff));
    }

    public double standardDeviation() {
        return standardDeviation;
    }

    public Map<Partner, Double> percentageErrorView() {
        return percentErrorMap;
    }

    public double errorFor(Partner p) {
        return MapOps.unbox(percentErrorMap.get(p));
    }
}

File: modules/engine/src/main/scala/api/queue/PartnerFairness.scala
package edu.gemini.tac.qengine.api.queue

import edu.gemini.tac.qengine.util.Time
import edu.gemini.tac.qengine.ctx.Partner

class PartnerFairness(remainingTime: Partner => Time, availableTime: Partner => Time, partners: List[Partner]) {

  private def hours(time: Time): Double = time.toHours.value

  private def calcErrorPercent(p: Partner): Double =
    hours(availableTime(p)) match {
      case 0.0   => 0.0
      case avail => - hours(remainingTime(p))/avail * 100
    }

  /**
   * A map from Partner to the percentage of error from the specified partner
   * allocation.  Negative numbers indicate under allocation and positive
   * numbers over allocation.
   */
  val errorPercent: Map[Partner, Double] =
    partners.map(p => p -> calcErrorPercent(p)).toMap

  private def avg(f: Partner => Double): Double =
    (0.0/:partners) { _ + f(_) } / partners.length

  val minErrorPercent: Double  = errorPercent.values.min
  val maxErrorPercent: Double  = errorPercent.values.min
  val meanErrorPercent: Double = avg(errorPercent)

  val medianErrorPercent: Double = errorPercent.values.toList.sorted match {
    case lst if lst.size % 2 == 1 => lst(lst.size / 2)
    case lst => {
      val second = lst.size / 2
      val first  = second - 1
      (lst(first) + lst(second))/2
    }
  }

  private def squareDiffFromMean(p: Partner): Double =
    scala.math.pow(errorPercent(p) - meanErrorPercent, 2.0)

  /**
   * The standard deviation of all the remaining time percentages.  The lower
   * the standard deviation, the more fair the proposal queue.
   */
  val standardDeviation: Double = scala.math.sqrt(avg(squareDiffFromMean))
}